rust/src/nasl/builtin/cert/mod.rs

```diff
@@ -0,0 +1,415 @@
+        "1.3.132.1.12" => "id-ecDH",
+        "1.2.840.10045.2.13" => "id-ecMQV",
+        "1.2.840.113549.1.1.10" => "id-RSASSA-PSS",
+        "1.2.840.113549.1.1.11" => "sha256WithRSAEncryption",
+        "1.2.840.113549.1.1.12" => "sha384WithRSAEncryption",
+        "1.2.840.113549.1.1.13" => "sha512WithRSAEncryption",
+        "1.2.840.113549.1.1.14" => "sha224WithRSAEncryption",
+        "1.2.840.113549.1.1.8" => "id-mgf1",
+        "1.2.840.113549.2.2" => "md2",
+        "1.2.840.113549.2.4" => "md4",
+        "1.2.840.113549.2.5" => "md5",
+        "1.2.840.113549.1.1.1" => "rsaEncryption",
+        "1.2.840.113549.1.1.2" => "md2WithRSAEncryption",
+        "1.2.840.113549.1.1.3" => "md4WithRSAEncryption",
+        "1.2.840.113549.1.1.4" => "md5WithRSAEncryption",
+        "1.2.840.113549.1.1.6" => "rsaOAEPEncryptionSET",
+        "1.2.840.10045.3.1.1" => "secp192r1",
+        "1.3.132.0.1" => "sect163k1",
+        "1.3.132.0.15" => "sect163r2",
+        "1.3.132.0.33" => "secp224r1",
+        "1.3.132.0.26" => "sect233k1",
+        "1.3.132.0.27" => "sect233r1",
+        "1.2.840.10045.3.1.7" => "secp256r1",
+        "1.3.132.0.16" => "sect283k1",
+        "1.3.132.0.17" => "sect283r1",
+        "1.3.132.0.34" => "secp384r1",
+        "1.3.132.0.36" => "sect409k1",
+        "1.3.132.0.37" => "sect409r1",
+        "1.3.132.0.35" => "sect521r1",
+        "1.3.132.0.38" => "sect571k1",
+        "1.3.132.0.39" => "sect571r1",
+        "2.16.840.1.101.3.4.3.1" => "id-dsa-with-sha224",
+        "2.16.840.1.101.3.4.3.2" => "id-dsa-with-sha256",
+        "2.16.840.1.101.3.4.2.1" => "sha256",
+        "2.16.840.1.101.3.4.2.2" => "sha384",
+        "2.16.840.1.101.3.4.2.3" => "sha512",
+        "2.16.840.1.101.3.4.2.4" => "sha224",
+        _ => "unknown",
+    }
+    .to_string()
+}
+
+fn pub_key_alg_oid_to_name(name: &str) -> String {
+    match name {
+        "1.2.840.113549.1.1.1" => "RSA",
+        "2.5.8.1.1" => "RSA (X.509)",
+        "1.2.840.113549.1.1.4" => "RSA (MD5)",
+        "1.2.840.113549.1.1.5" => "RSA (SHA1)",
+        "1.2.840.10040.4.1" => "DSA",
+        "1.2.643.2.2.19" => "GOST R 34.10-2001",
+        "1.2.643.2.2.20" => "GOST R 34.10-94",
+        "1.2.840.10045.2.1" => "EC",
+        _ => "unknown",
+    }
+    .to_string()
+}
+
+fn subject_oid_to_name(oid: &str) -> String {
+    match oid {
+        "2.5.4.6" => "C",
+        "2.5.4.8" => "ST",
+        "2.5.4.7" => "L",
+        "2.5.4.10" => "O",
+        "2.5.4.3" => "CN",
+        "2.5.4.11" => "OU",
+        "2.5.4.12" => "T",
+        "2.5.4.42" => "GN",
+        "2.5.4.43" => "I",
+        "2.5.4.4" => "SN",
+        _ => oid,
+    }
+    .to_string()
+}
+
+/// This structure holds a list of certificates. The entries of the list are
+/// Optional to allow for the removal of certificates. The closed list holds
+/// the indexes of the removed certificates.
+#[derive(Default)]
+struct Handles {
+    certs: HashMap<usize, X509Certificate>,
```

Comment:
why hashmap instead of a Vector? Next just seems to be an index anyway?

Reply:
Both have advantages and disadvantages. In the end the HashMap is more space efficient, as I do not have to store `None` variants for freed Certificates. For space efficiency within a Vector I also had to use an Vector of indices for free Certificates, so the variant with the next variable is also more efficient. In the end it will probably not make a huge difference.

Reply:
How can a HashMap of usize be more space efficient than an indexed vector?